library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileInputOperator.java

```diff
@@ -569,7 +616,9 @@ protected void replay(long windowId)
               if (inputStream != null) {
                 closeFile(inputStream);
               }
-              processedFiles.add(recoveryEntry.file);
+              synchronized (processedFiles.getClass()) {
```

Comment:
I'm leaning towards @tushargosavi 's position: Explicit synchronization is error prone and is better avoided if possible. `ConcurrentHashMap` has a default constructor and can simply be used as a set by using a single fixed value (e.g. `Boolean.TRUE`) for keys that are present in the set. Or you could even explicitly define a `ConcurrentHashSet` class as shown by response 4 at:
http://stackoverflow.com/questions/6992608/why-there-is-no-concurrenthashset-against-concurrenthashmap